src/librustc/hir/lowering.rs

```diff
@@ -3363,9 +3410,9 @@ impl<'a> LoweringContext<'a> {
                 (generics, hir::TraitItemKind::Method(sig, hir::TraitMethod::Provided(body_id)))
             }
             TraitItemKind::Type(ref bounds, ref default) => {
-                let generics = self.lower_generics(&i.generics, ImplTraitContext::AssociatedTy);
+                let generics = self.lower_generics(&i.generics, ImplTraitContext::disallowed());
```

Comment:
This `disallowed` thing feels like a bit of a hack / makes me uncomfortable, but I don't have sufficient context to judge here.